---
title: 'PDX: Drug trails'
author: "Alistair Martin"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output:
  pdf_document:
    toc: yes
  html_notebook:
    toc: yes
    toc_float: yes
  html_document:
    toc: yes
    toc_float: yes
subtitle: GDC (Susana)
layout: page
---

# Load libraries, functions and data
```{r, message=F, warning=F, echo=F}
root.dir <- "~/OneDrive/projects/"
project.dir <- paste0(root.dir,"PDX/")
source(paste0(root.dir,"R.utils/RNASEQ.utils.R"))
source(paste0(project.dir,"src/load.PDX.R"))

rnaseq <- get.PDX()
meta <- load.meta()
meta <- subset.meta(meta,meta[,project=="drug.GDC"])
rnaseq <- subset.PDX(rnaseq, rnaseq$meta[,Sample.name] %in% meta[,Sample.name])
```

#QC

```{r}
read.depth.plots(rnaseq$meta)
```


We appear to have some very small libraries. Given how few samples we have, let's generate a table with read counts.

```{r}
rnaseq$meta[order(human.library.size),.(Sample.name,Pool,human.library.size)]
```

However, note that the samples in SLX-14086, the "failed" libraries, are the same as those in SLX-14085. Given this, the trail arms are fully present and we can perform the analysis. First, let us merge the lanes.

```{r}
#rnaseq <- subset.PDX(rnaseq, rnaseq$meta[,human.library.size>1E6])
rnaseq <- merge.PDX(rnaseq)
rnaseq$meta <- merge(rnaseq$meta,meta,sort=F)
rnaseq$meta[,merged.lanes:=factor(as.numeric(as.factor(merged.lanes)))]
```

Let's just double check the trail arms.

```{r}
table(rnaseq$meta[,Treatment])
```

# PCAs

```{r}
pca <- pca.run(rnaseq$human)
summary(pca)
pca <- cbind(rnaseq$meta,pca$x)
ggplot(pca) + aes(x=PC1,y=PC2,col=Treatment,shape=merged.lanes) + geom_point(size=5)
```

We appear to have a batch effect associated with the sequencing that we need to include in the statistical models. How does the pca change if we try accounting for it?

```{r}
y <- pca.prep(rnaseq$human)
y <- removeBatchEffect(y,design = model.matrix(~rnaseq$meta$Treatment), batch= rnaseq$meta$merged.lanes)

pca <- prcomp(t(y),scale=T)
summary(pca)
pca <- cbind(rnaseq$meta,pca$x)
ggplot(pca) + aes(x=PC1,y=PC2,col=Treatment,shape=merged.lanes) + geom_point(size=5)
```


That seems to work well. :-)

# Model: ~ Treatment + Flowcell

```{r}
sequencing <- rnaseq$meta[,merged.lanes]
treatment <- rnaseq$meta[,factor(Treatment,levels = c("control","GDC"))]
d <- model.matrix(~ 0 + treatment + sequencing)
d

cont.matrix <- matrix(c(-1,1,rep(0,dim(d)[2]-2)), dim(d)[2], 1)
colnames(cont.matrix) <- c("GDP.minus.control")
rownames(cont.matrix) <- colnames(d)
cont.matrix

#Perform DE
DE.data <- DE.run(rnaseq$human,d,2,cont.matrix = cont.matrix)

#Top DE genes table
DE <- topTable.annotated.DE(DE.data$efit)
if(nrow(DE)!=0){
  table(DE[,direction])
  DE[direction=="up.reg",.(direction,ENSEMBL,SYMBOL,adj.P.Val,logFC)]
  DE[direction=="down.reg",.(direction,ENSEMBL,SYMBOL,adj.P.Val,logFC)]

#Pheatmap w/ batch correction
  i <- match(DE$ENSEMBL,rownames(DE.data$v$E))
  y <- removeBatchEffect(DE.data$v$E[i,],design = model.matrix(~drug), batch= pdx)
  labels.col <- rep("",nrow(rnaseq$meta))
  labels.row <- DE$SYMBOL
  ann.col <- as.data.frame(rnaseq$meta[,.(pdx,drug)])
  rownames(ann.col) <- rnaseq$meta[,Sample.name]
  pheatmap(y,labels_col=labels.col,labels_row=labels.row,annotation_col = ann.col,scale="row")
}

#GSEA - c6
gsea <- GSEA.run(DE.data$v,d,cont.matrix[,1],Hs.c6,0.05)
gsea[Direction=="Up",.(rn,NGenes,Direction,FDR)]
gsea[Direction=="Down",.(rn,NGenes,Direction,FDR)]

# ssGSEA
ssGSEA.data <- ssGSEA.run(DE.data$v$E,Hs.c2,d,cont.matrix=cont.matrix)

#Top DE gene.sets table
ssGSEA <- topTable.annotated.ssGSEA(ssGSEA.data$efit)
if(nrow(ssGSEA)!=0){
  ssGSEA[direction=="up.reg",.(direction,gene.set,adj.P.Val,logFC)]
  ssGSEA[direction=="down.reg",.(direction,gene.set,adj.P.Val,logFC)]
  
  #Pheatmap w/ batch correction
  i <- match(ssGSEA$gene.set,rownames(ssGSEA.data$w))
  y <- removeBatchEffect(ssGSEA.data$w[i,],design = model.matrix(~drug), batch=pdx)
  labels.col <- rep("",nrow(rnaseq$meta))
  labels.row <- ssGSEA$gene.set
  ann.col <- as.data.frame(rnaseq$meta[,.(pdx,drug)])
  rownames(ann.col) <- rnaseq$meta[,Sample.name]
  pheatmap(y,labels_col=labels.col,labels_row=labels.row,annotation_col = ann.col,scale="row")
}
```

Sadly, we do not see have anything reported as DE. This is most likely due to the batch effect generated by the sequencing being performed across two different flowcells. With this compounding factor, we will always struggle to obtain anything relevant from a standard differential expression analysis of six samples. My personal suggestion would be to repool and resequecing everything together so as to remove this.

## Specific genes

The below plots show the normalised expression for the genes of interest across the two conditions. We have shifted the expressions in an attempt to account for the batch effect.

```{r}
genes <- c("PTEN","FOXM1","LDHA","HK2","MYC")
genes.ensembl <- mapIds(org.Hs.eg.db,keys=genes,column = "ENSEMBL",keytype = "SYMBOL")
i <- match(genes.ensembl,rownames(DE.data$v$E))
x <- t(removeBatchEffect(DE.data$v$E[i,],design = model.matrix(~treatment), batch=sequencing))
colnames(x) <- genes 
x <- cbind(rnaseq$meta[,.(Treatment,merged.lanes)],x)
x <- melt(x,id.vars = c("Treatment","merged.lanes"))
ggplot(x) + aes(x=Treatment,y=value) + geom_point() + facet_wrap(~variable,scales = "free_y")
```

## Specific pathways

The below plots show the normalised expression for the pathways of interest across the two conditions. We have shifted the expressions in an attempt to account for the batch effect.

```{r}
i <- grep("PI3K",rownames(ssGSEA.data$w))
x <- t(removeBatchEffect(ssGSEA.data$w[i,],design = model.matrix(~treatment), batch=sequencing))
x <- cbind(rnaseq$meta[,.(Treatment,merged.lanes)],x)
x <- melt(x,id.vars = c("Treatment","merged.lanes"))
ggplot(x) + aes(x=variable,y=value,colour=Treatment) + geom_point() + coord_flip() + theme(axis.text.y = element_text(size=6))
```

