---
title: 'PDX: COX2-Stroma pathway'
author: "Alistair Martin"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output:
  html_document:
    toc: yes
    toc_float: yes
    code_folding: hide
layout: page
---


#Load libraries and data

```{r, message=FALSE, warning=FALSE}
root.dir <- "~/OneDrive/projects/"
source(paste0(root.dir,"R.utils/RNASEQ.utils.R"))

project.dir <- paste0(root.dir,"PDX/")
source(paste0(project.dir,"src/load.PDX.R"))
rnaseq <- get.PDX()
meta <- load.meta()

meta <- subset.meta(meta,meta[,project=="characterisation"])
rnaseq <- subset.PDX(rnaseq, rnaseq$meta[,Sample.name] %in% meta[,Sample.name])
rnaseq <- merge.PDX(rnaseq)
rnaseq$meta[,merged.lanes:=as.numeric(as.factor(merged.lanes))]

rnaseq$meta <- merge(rnaseq$meta,meta,sort=F)
rnaseq <- subset.PDX(rnaseq, rnaseq$meta[,sample.type == "xenograft"])
rnaseq <- subset.PDX(rnaseq, rnaseq$meta[,human.library.size > 1e6])
rnaseq <- subset.PDX(rnaseq, rnaseq$meta[,mouse.library.size/(human.library.size+mouse.library.size) < .8])
```

```{r}
x <-  norm.count.matrix(rnaseq$human)
#x <- removeBatchEffect(x,batch=rnaseq$meta$merged.lanes)
i <- which(rownames(x) == "ENSG00000073756")
COX2.human.exp <- x[i,]
pca.human <- pca.run(x,normalise = F)


ggpairs(cbind(pca.human,rnaseq$meta,COX2.human.exp),columns = 1:3,mapping = aes(color=factor(merged.lanes)),title="Coloured by sequencing runs")
ggpairs(cbind(pca.human,rnaseq$meta,COX2.human.exp),columns = 1:3,mapping = aes(color=factor(patient)),title="Coloured by patient")
ggpairs(cbind(pca.human,rnaseq$meta,COX2.human.exp),columns = 1:3,mapping = aes(color=cut(mouse.library.size/(human.library.size+mouse.library.size),quantile(mouse.library.size/(human.library.size+mouse.library.size)))),title="Coloured by mouse fraction")
ggpairs(cbind(pca.human,rnaseq$meta,COX2.human.exp),columns = 1:3,mapping = aes(color=cut(COX2.human.exp,quantile(COX2.human.exp))),title="Coloured by COX2.human.exp")
ggpairs(cbind(pca.human,rnaseq$meta,COX2.human.exp)[mouse.library.size/(human.library.size+mouse.library.size)>0.01],columns = 1:3,mapping = aes(color=cut(COX2.human.exp,quantile(COX2.human.exp))),title="Coloured by COX2.human.exp (subsetted by stroma fraction)")
```

Provided there is some (mouse) stroma present in the sample, there appears to be some clustering related to the COX2 expression, specifically from dimension PC2. However, subsetting to only those samples with adequete amounts of stroma, we find the clustering not as strong as it appeared before. 

# COX2 Signatures w.r.t human reads

```{r}
PoI <- list(
  "COX2_UP_UP" = "IL6 IL1B IL1A CXCL1 CXCL2 CSF3 CCL2 PTGS2 VEGFA",
  "COX2_UP_DN" = "IL12B CXCL9 CXCL10 CXCL11 TBX21 CCL5 IFNG GZMB PRF1 IL12A STAT1 IFIT1 IFIT2"
)
PoI <- sapply(PoI, strsplit, split=" ")
PoI <- lapply(PoI, function(x) data.table(biomaRt::select(org.Hs.eg.db,keys=x,keytype = "SYMBOL",columns = c("ENSEMBL","ENTREZID","SYMBOL"))))

x <- norm.count.matrix(rnaseq$human[rowSums(rnaseq$human)!=0,])
i <- which(rownames(x) == "ENSG00000073756")
COX2.human.exp <- x[i,]

y <- gsva(x,lapply(PoI,"[[","ENSEMBL"),method="ssgsea")

x <- cbind(rnaseq$meta,COX2.human.exp,t(y))
x[,mouse.frac:=mouse.library.size/(human.library.size+mouse.library.size)]

ggpairs(x,columns = c("COX2.human.exp","COX2_UP_UP","COX2_UP_DN"))

summary(x[,lm(COX2.human.exp ~ merged.lanes + mouse.frac)])
summary(x[,lm(COX2.human.exp ~ COX2_UP_UP)])
summary(x[,lm(COX2.human.exp ~ COX2_UP_DN)])
```


COX2.human exp is found to positively correlate to the pathway score of COX2_UP_UP within the human reads. 

# COX2 Signatures w.r.t mouse reads

## All PDXs

```{r}
firstup <- function(x){
  x <- tolower(x)
  substr(x, 1, 1) <- toupper(substr(x, 1, 1))
  return(x)
}
PoI <- list(
  "COX2_UP_UP" = "IL6 IL1B IL1A CXCL1 CXCL2 CSF3 CCL2 Ptgs2 Vegfa",
  "COX2_UP_DN" = "IL12B CXCL9 CXCL10 CXCL11 TBX21 CCL5 IFNG GZMB PRF1 IL12A STAT1 IFIT1 IFIT2"
)
PoI <- sapply(PoI, strsplit, split=" ")
PoI <- lapply(PoI, function(p) sapply(p,firstup))
PoI <- lapply(PoI, function(x) data.table(biomaRt::select(org.Mm.eg.db,keys=x,keytype = "SYMBOL",columns = c("ENSEMBL","ENTREZID","SYMBOL"))))

x <- norm.count.matrix(rnaseq$human[rowSums(rnaseq$human)!=0,])
i <- which(rownames(x) == "ENSG00000073756")
COX2.human.exp <- x[i,]

y <- norm.count.matrix(rnaseq$mouse[rowSums(rnaseq$mouse)!=0,])
y <- gsva(y,lapply(PoI,"[[","ENSEMBL"),method="ssgsea")

x <- cbind(rnaseq$meta,COX2.human.exp,t(y))
x[,mouse.frac:=mouse.library.size/(human.library.size+mouse.library.size)]

ggpairs(x,columns = c("COX2.human.exp","COX2_UP_UP","COX2_UP_DN","mouse.library.size"))

summary(x[,lm(COX2.human.exp ~ merged.lanes + mouse.frac + log(mouse.library.size))])
summary(x[,lm(COX2_UP_UP ~ COX2.human.exp + merged.lanes + mouse.frac + log(mouse.library.size))])
summary(x[,lm(COX2_UP_DN ~ COX2.human.exp + merged.lanes + mouse.frac + log(mouse.library.size))])
```


We find that the pathway COX2_UP_DN within the mouse compoenent is heavily correlated to the mouse.library size. However, our cohort contains samples with as few mouse reads as 10s of thousands. Conditioning on this however, we find that COX2.human exp is found to positively correlated to the pathway score of COX2_UP_DN within the mouse reads. This is the opposite of what is expected.

## 1E5 min. mouse reads

```{r}
firstup <- function(x){
  x <- tolower(x)
  substr(x, 1, 1) <- toupper(substr(x, 1, 1))
  return(x)
}
PoI <- list(
  "COX2_UP_UP" = "IL6 IL1B IL1A CXCL1 CXCL2 CSF3 CCL2 Ptgs2 Vegfa",
  "COX2_UP_DN" = "IL12B CXCL9 CXCL10 CXCL11 TBX21 CCL5 IFNG GZMB PRF1 IL12A STAT1 IFIT1 IFIT2"
)
PoI <- sapply(PoI, strsplit, split=" ")
PoI <- lapply(PoI, function(p) sapply(p,firstup))
PoI <- lapply(PoI, function(x) data.table(biomaRt::select(org.Mm.eg.db,keys=x,keytype = "SYMBOL",columns = c("ENSEMBL","ENTREZID","SYMBOL"))))

j <- rnaseq$meta[,mouse.library.size>1E5]

x <- norm.count.matrix(rnaseq$human[rowSums(rnaseq$human)!=0,j])
i <- which(rownames(x) == "ENSG00000073756")
COX2.human.exp <- x[i,]

y <- norm.count.matrix(rnaseq$mouse[rowSums(rnaseq$mouse)!=0,j])
y <- gsva(y,lapply(PoI,"[[","ENSEMBL"),method="ssgsea")

x <- cbind(rnaseq$meta,COX2.human.exp,t(y))
x[,mouse.frac:=mouse.library.size/(human.library.size+mouse.library.size)]

ggpairs(x,columns = c("COX2.human.exp","COX2_UP_UP","COX2_UP_DN","mouse.library.size"))

summary(x[,lm(COX2.human.exp ~ merged.lanes + mouse.frac + log(mouse.library.size))])
summary(x[,lm(COX2_UP_UP ~ COX2.human.exp + merged.lanes + mouse.frac + log(mouse.library.size))])
summary(x[,lm(COX2_UP_DN ~ COX2.human.exp + merged.lanes + mouse.frac + log(mouse.library.size))])
```

Removing the the samples with low numbers of mouse reads removes the dependency of COX2_UP_DN on the mouse.library.size. However, we still find that COX2.human exp is found to positively correlated to the pathway score of COX2_UP_DN within the mouse reads. Again, this is the opposite of what is expected.