---
title: 'PDX: PAR1040'
author: "Alistair Martin"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output:
  html_document:
    toc: yes
    toc_float: yes
    code_folding: hide
layout: page
---

#Load libraries and data

```{r, message=FALSE, warning=FALSE}
root.dir <- "~/OneDrive/projects/"
source(paste0(root.dir,"R.utils/RNASEQ.utils.R"))
library(UpSetR)

project.dir <- paste0(root.dir,"PDX/")
source(paste0(project.dir,"src/load.PDX.R"))

results.dir <- paste0(project.dir,"results/PARTNER/")
invisible(lapply(grep("dir$",ls(),value = T), function(dname) dir.create(eval(as.name(dname)),recursive = T,showWarnings = F)))

meta <- load.meta()
#meta <- subset.meta(meta,meta[,project %in% c("PAR1040","PAR1006")])
meta <- subset.meta(meta,meta[,project %in% c("PAR1040","PAR1006","PAR1022")])

rnaseq <- get.PDX()
rnaseq <- subset.PDX(rnaseq, rnaseq$meta[,Sample.name] %in% meta[,Sample.name])
```

```{R}
read.depth.plots(rnaseq$meta)
```


```{R}
rnaseq <- merge.PDX(rnaseq)
rnaseq$meta[,merged.lanes:=as.numeric(as.factor(merged.lanes))]
rnaseq$meta <- merge(rnaseq$meta,meta,sort=F)
```

```{r}
print(rnaseq$meta[human.library.size < 1e6,Sample.name])
rnaseq <- subset.PDX(rnaseq, rnaseq$meta[,human.library.size > 1e6])
```

Removed "PAR1006-T1-P01-17.3840-T1-FF1" & "PAR1040-T1-P01-17.45011-T1-FF1"

```{r}
rnaseq <- subset.PDX(rnaseq, -rnaseq$meta[,which(Trial == "ABT008B")])
```

```{r}
pca.human <- pca.run(rnaseq$human$data,rename = T)
pca.human <- cbind(pca.human,rnaseq$meta)
ggpairs(pca.human,columns = 1:3,mapping = aes(color=factor(project)),title="Coloured by cohort")
```

The two pdx models seperate cleanly, with this defining the majority of the variation.

```{r}
rnaseq$meta[,Trial:=factor(Trial,levels=c("ABT008","ABT008A"))]
rnaseq$meta[project=="PAR1006",Cohort:=make.names(tstrsplit(Cohort," - ")[[2]])]
rnaseq$meta[Cohort=="Olaparib...AZD1775",Cohort:="Olap.AZD1775"]
rnaseq$meta[Cohort=="Olap+AZD1775",Cohort:="Olap.AZD1775"]

rnaseq$meta[,Cohort:=factor(Cohort)]
rnaseq$meta[,Cohort:=relevel(Cohort,"Untreated")]

rnaseq$meta[Sampling=="end",Sampling:="PT"]
rnaseq$meta[Sampling=="mid",Sampling:="T"]
rnaseq$meta[Sampling=="HC sch1",Sampling:="premature"]
```

```{r}
rnaseq$meta[,.N,.(project,Cohort,Sampling,Trial)][order(Cohort,project)]
```

I dislike testing the midpoints (T) due to lack of replicates. Let's focus on the post-treated samples.

```{r,eval=F}
library(readxl)

fnames <- c("~/Desktop/PAR1022_PAR1040_RNA_submission.xlsx","~/Desktop/30102018_CClab_RNA_PLATE2.xlsx")

x1 <-  data.table(read_xlsx("~/Desktop/PAR1022_PAR1040_RNA_submission.xlsx",sheet = 1,skip = 21))[,-1]
x1 <- x1[,.(Sample.name=Name,Row,Column,Submission=1)]

x2 <- data.table(read_xlsx("~/Desktop/30102018_CClab_RNA_PLATE2.xlsx",sheet = 1,skip = 27))[-(1:2),-1]
names(x2) <- make.names(names(x2))
x2 <- x2[,.(Sample.name=Sample.Name,Row=UDF.Row,Column=UDF.Column,Submission=2)]

x <- rbind(x1,x2)

rnaseq$meta <- merge(rnaseq$meta,x,by="Sample.name",all.x=T,sort=F)
```


```{r}
PAR1040 <- subset.PDX(rnaseq,rnaseq$meta[,project=="PAR1040"])
PAR1006 <- subset.PDX(rnaseq,rnaseq$meta[,project=="PAR1006"])
```

# PAR1040

## PCAs w.r.t. batch

```{r}
pca.human <- pca.run(PAR1040$human$data,rename = T)
pca.human <- cbind(pca.human,PAR1040$meta)

ggpairs(pca.human,columns = 1:3,mapping = aes(color=factor(Cohort)),title="Coloured by cohort")
ggpairs(pca.human,columns = 1:3,mapping = aes(color=cut(mouse.library.size/(human.library.size+mouse.library.size),quantile(mouse.library.size/(human.library.size+mouse.library.size)))),title="Coloured by mouse fraction")
```

## PCAs w.r.t cohort

```{r}
for(coh in unique(PAR1040$meta[,Cohort])){
  x <- subset.PDX(PAR1040, PAR1040$meta[,Cohort %in% c(coh,"Untreated")])
  pca.human <- pca.run(x$human$data,rename = T)
  pca.human <- cbind(pca.human,x$meta)
  print(suppressWarnings(ggpairs(pca.human,columns = 1:3,mapping = aes(color=factor(Cohort),shape=factor(Sampling)),title=paste0(coh,": Coloured by cohort"),)))
}
```

## Model: ~ 0 + cohort

```{r}
results.dir <- paste0(project.dir,"results/PARTNER/PAR1040/")
invisible(lapply(grep("dir$",ls(),value = T), function(dname) dir.create(eval(as.name(dname)),recursive = T,showWarnings = F)))

x <- subset.PDX(PAR1040, PAR1040$meta[,Cohort %in% c("Untreated","Olaparib","AZD1775","Olap.AZD1775") & Sampling=="PT"])
cohort <- droplevels(x$meta[,Cohort])
d <- model.matrix(~ 0 + cohort)

n <- length(levels(cohort)) - 1
cont <- rbind(-1,diag(n))
colnames(cont) <- levels(cohort)[-1]
rownames(cont) <- colnames(d)
cont

DE.data <- DE.run(x$human$data,d,cont.matrix=cont)

#Tables of DE genes
for(i in 1:n){
  DE <- topTable.annotated.DE(DE.data$efit,coef = i)
  fwrite(DE,paste0(results.dir,"DE.genes.",colnames(cont)[i],".csv"))
}

#Overlap of DE genes
DE <- decideTest.annotated.DE(DE.data$efit,p.val = 0.05)
fwrite(DE[.N:1],paste0(results.dir,"DE.genes.combined.csv"))

#Overlap plot
if(nrow(DE)){
  x <- rbind(
    cbind(DE[,ifelse(.SD[,2:(n+1)]>0,1,0)],data.table(reg="UP")),
    cbind(DE[,ifelse(.SD[,2:(n+1)]<0,1,0)],data.table(reg="DOWN"))
  )
  p <- upset(
    x,
    queries = list(list(query=elements,params=list("reg","UP"),color="red",active=T)),
    order.by = "freq",
    nsets = n,
    nintersects = 20
  )
  png(paste0(results.dir,"DE.genes.combined.png"),width = 1500,height = 1000,res=300)
  p
  dev.off()
}

#Gene set 
hm.gs <- load_gene_set(paste0(root.dir,"shared.data/gene.lists/h.all.v6.2.entrez.gmt"))
x <- lapply(1:n,function(i) GSEA.run(DE.data$v,d,cont[,i],hm.gs,0.05,igc=0.05))
names(x) <- colnames(cont)
for(i in 1:n) fwrite(x[[i]],paste0(results.dir,"GSEA.hallmark.",colnames(cont)[i],".csv"))

#Gene set overlap
x <- rbindlist(x,idcol = "contrast")
x[,rank:=1:.N,contrast]
x <-  rbind(
  cbind(dcast(x[Direction=="Up"],pathway~contrast,value.var = "rank"),data.table(reg="Up")),
  cbind(dcast(x[Direction=="Down"],pathway~contrast,value.var = "rank"),data.table(reg="Down")),
  fill=TRUE
)
x <- cbind(x[,-grep("reg",colnames(x)),with=F],x[,.(reg)])
x[,avg.rank:=rowMeans(.SD[,2:(ncol(x)-1)],na.rm = T)]
fwrite(x[order(avg.rank)],paste0(results.dir,"GSEA.hallmark.combined.csv"))

#Overlap plot
x <- cbind(x[,.(reg)],x[,ifelse(!is.na(.SD[,2:(ncol(x)-2)]),1,0)])
p <- upset(x, queries = list(list(query=elements,params=list("reg","Up"),color="red",active=T)), order.by = "freq",  nsets = n)
png(paste0(results.dir,"GSEA.hallmark.combined.png"),width = 1500,height = 1000,res=300)
p
dev.off()


# scRNAseq cluster markers vs biobank
scRNAseq.dir <- paste0(root.dir,"PDX.scRNAseq/results/human/PAR1040.combined/")
fnames <- grep("cluster.\\d.marker.genes.csv",list.files(scRNAseq.dir),value = T)
scRNAseq.genes.par1040 <- lapply(fnames, function(fname){
  x <- fread(paste0(scRNAseq.dir,fname))
  x[,gene:=gsub("hg19-","",gene)]
  x[,gene.entrez:=mapIds(org.Hs.eg.db,keys=gene,keytype="SYMBOL",column="ENTREZID",multiVals="first")]
  y <- x[,grep("logFC",names(x)),with=F]>0
  x$direction <- ifelse(apply(y,1,all),"UP",ifelse(apply(!y,1,all),"DOWN","MIXED"))
  list(
    "UP"=na.omit(x[direction=="UP",gene.entrez]),
    "DOWN"=na.omit(x[direction=="DOWN",gene.entrez])
  )
})
for(i in seq_along(scRNAseq.genes.par1040)) names(scRNAseq.genes.par1040[[i]]) <- paste0("cluster.",i-1,".",names(scRNAseq.genes.par1040[[i]]))
scRNAseq.genes.par1040 <- unlist(scRNAseq.genes.par1040,recursive = F)
x <- lapply(1:n,function(i) GSEA.run(DE.data$v,d,cont[,i],scRNAseq.genes.par1040,0.05,igc=0.05))
names(x) <- colnames(cont)

for(i in 1:n) fwrite(x[[i]],paste0(results.dir,"GSEA.scRNAseq.",colnames(cont)[i],".csv"))

#Gene set overlap
x <- rbindlist(x,idcol = "contrast")
x[,rank:=1:.N,contrast]
x <-  rbind(
  cbind(dcast(x[Direction=="Up"],pathway~contrast,value.var = "rank"),data.table(reg="Up")),
  cbind(dcast(x[Direction=="Down"],pathway~contrast,value.var = "rank"),data.table(reg="Down")),
  fill=TRUE
)
x <- cbind(x[,-grep("reg",colnames(x)),with=F],x[,.(reg)])
x[,avg.rank:=rowMeans(.SD[,2:(ncol(x)-1)],na.rm = T)]
fwrite(x[order(avg.rank)],paste0(results.dir,"GSEA.scRNAseq.combined.csv"))

#Overlap plot
x <- cbind(x[,.(reg)],x[,ifelse(!is.na(.SD[,2:(ncol(x)-2)]),1,0)])
p <- upset(x, queries = list(list(query=elements,params=list("reg","Up"),color="red",active=T)), order.by = "freq",  nsets = n)
png(paste0(results.dir,"GSEA.scRNAseq.combined.png"),width = 1500,height = 1000,res=300)
p
dev.off()
```


# PAR1006

## PCAs w.r.t. batch

```{r}
pca.human <- pca.run(PAR1006$human$data,rename = T)
pca.human <- cbind(pca.human,PAR1006$meta)

ggpairs(pca.human,columns = 1:3,mapping = aes(color=factor(merged.lanes)),title="Coloured by sequencing runs")
ggpairs(pca.human,columns = 1:3,mapping = aes(color=factor(Cohort)),title="Coloured by cohort")
ggpairs(pca.human,columns = 1:3,mapping = aes(color=cut(mouse.library.size/(human.library.size+mouse.library.size),quantile(mouse.library.size/(human.library.size+mouse.library.size)))),title="Coloured by mouse fraction")
```

No major correction needed for sequencing runs which is a good sign.

## PCAs w.r.t cohort

```{r}
for(coh in unique(PAR1006$meta[,Cohort])){
  x <- subset.PDX(PAR1006, PAR1006$meta[,Cohort %in% c(coh,"Untreated")])
  pca.human <- pca.run(x$human$data,rename = T)
  pca.human <- cbind(pca.human,x$meta)
  print(suppressWarnings(ggpairs(pca.human,columns = 1:3,mapping = aes(color=factor(Cohort),shape=factor(Sampling)),title=paste0(coh,": Coloured by cohort"),)))
}
```

## Model: ~ 0 + cohort + trial + sequencing + mouse

```{r}
results.dir <- paste0(project.dir,"results/PARTNER/PAR1006/")
invisible(lapply(grep("dir$",ls(),value = T), function(dname) dir.create(eval(as.name(dname)),recursive = T,showWarnings = F)))

x <- subset.PDX(PAR1006, PAR1006$meta[,Cohort %in% c("Untreated","Olaparib","AZD1775","Olap.AZD1775") & Sampling=="PT"])
cohort <- droplevels(x$meta[,Cohort])
trial <- x$meta$Trial
mouse <- x$meta$Mouse
sequencing <- factor(x$meta$merged.lanes)
d <- model.matrix(~ 0 + cohort + trial + sequencing + mouse)
d <- d[,colSums(d)>1]

n <- length(levels(cohort)) - 1
cont <- rbind(-1,diag(n))
colnames(cont) <- levels(cohort)[-1]
for(i in 1:(ncol(d)-nrow(cont))) cont <- rbind(cont,0)
rownames(cont) <- colnames(d)
cont

DE.data <- DE.run(x$human$data,d,cont.matrix=cont)

#Tables of DE genes
for(i in 1:n){
  DE <- topTable.annotated.DE(DE.data$efit,coef = i)
  fwrite(DE,paste0(results.dir,"DE.genes.",colnames(cont)[i],".csv"))
}

#Overlap of DE genes
DE <- decideTest.annotated.DE(DE.data$efit,p.val = 0.05)
fwrite(DE[.N:1],paste0(results.dir,"DE.genes.combined.csv"))

#Overlap plot
if(nrow(DE)){
  x <- rbind(
    cbind(DE[,ifelse(.SD[,2:(n+1)]>0,1,0)],data.table(reg="UP")),
    cbind(DE[,ifelse(.SD[,2:(n+1)]<0,1,0)],data.table(reg="DOWN"))
  )
  p <- upset(
    x,
    queries = list(list(query=elements,params=list("reg","UP"),color="red",active=T)),
    order.by = "freq",
    nsets = n,
    nintersects = 20
  )
  png(paste0(results.dir,"DE.genes.combined.png"),width = 1500,height = 1000,res=300)
  p
  dev.off()
}

#Gene set 
hm.gs <- load_gene_set(paste0(root.dir,"shared.data/gene.lists/h.all.v6.2.entrez.gmt"))
x <- lapply(1:n,function(i) GSEA.run(DE.data$v,d,cont[,i],hm.gs,0.05,igc=0.05))
names(x) <- colnames(cont)
for(i in 1:n) fwrite(x[[i]],paste0(results.dir,"GSEA.hallmark.",colnames(cont)[i],".csv"))

#Gene set overlap
x <- rbindlist(x,idcol = "contrast")
x[,rank:=1:.N,contrast]
x <-  rbind(
  cbind(dcast(x[Direction=="Up"],pathway~contrast,value.var = "rank"),data.table(reg="Up")),
  cbind(dcast(x[Direction=="Down"],pathway~contrast,value.var = "rank"),data.table(reg="Down")),
  fill=TRUE
)
x <- cbind(x[,-grep("reg",colnames(x)),with=F],x[,.(reg)])
x[,avg.rank:=rowMeans(.SD[,2:(ncol(x)-1)],na.rm = T)]
fwrite(x[order(avg.rank)],paste0(results.dir,"GSEA.hallmark.combined.csv"))

#Overlap plot
x <- cbind(x[,.(reg)],x[,ifelse(!is.na(.SD[,2:(ncol(x)-2)]),1,0)])
p <- upset(x, queries = list(list(query=elements,params=list("reg","Up"),color="red",active=T)), order.by = "freq",  nsets = n)
png(paste0(results.dir,"GSEA.hallmark.combined.png"),width = 1500,height = 1000,res=300)
p
dev.off()


# scRNAseq cluster markers vs biobank
scRNAseq.dir <- paste0(root.dir,"PDX.scRNAseq/results/human/PAR1006.combined/")
fnames <- grep("cluster.\\d.marker.genes.csv",list.files(scRNAseq.dir),value = T)
scRNAseq.genes.par1006 <- lapply(fnames, function(fname){
  x <- fread(paste0(scRNAseq.dir,fname))
  x[,gene:=gsub("hg19-","",gene)]
  x[,gene.entrez:=mapIds(org.Hs.eg.db,keys=gene,keytype="SYMBOL",column="ENTREZID",multiVals="first")]
  y <- x[,grep("logFC",names(x)),with=F]>0
  x$direction <- ifelse(apply(y,1,all),"UP",ifelse(apply(!y,1,all),"DOWN","MIXED"))
  list(
    "UP"=na.omit(x[direction=="UP",gene.entrez]),
    "DOWN"=na.omit(x[direction=="DOWN",gene.entrez])
  )
})
for(i in seq_along(scRNAseq.genes.par1006)) names(scRNAseq.genes.par1006[[i]]) <- paste0("cluster.",i-1,".",names(scRNAseq.genes.par1006[[i]]))
scRNAseq.genes.par1006 <- unlist(scRNAseq.genes.par1006,recursive = F)
x <- lapply(1:n,function(i) GSEA.run(DE.data$v,d,cont[,i],scRNAseq.genes.par1006,0.05,igc=0.05))
names(x) <- colnames(cont)

for(i in 1:n) fwrite(x[[i]],paste0(results.dir,"GSEA.scRNAseq.",colnames(cont)[i],".csv"))

#Gene set overlap
x <- rbindlist(x,idcol = "contrast")
x[,rank:=1:.N,contrast]
x <-  rbind(
  cbind(dcast(x[Direction=="Up"],pathway~contrast,value.var = "rank"),data.table(reg="Up")),
  cbind(dcast(x[Direction=="Down"],pathway~contrast,value.var = "rank"),data.table(reg="Down")),
  fill=TRUE
)
x <- cbind(x[,-grep("reg",colnames(x)),with=F],x[,.(reg)])
x[,avg.rank:=rowMeans(.SD[,2:(ncol(x)-1)],na.rm = T)]
fwrite(x[order(avg.rank)],paste0(results.dir,"GSEA.scRNAseq.combined.csv"))

#Overlap plot
x <- cbind(x[,.(reg)],x[,ifelse(!is.na(.SD[,2:(ncol(x)-2)]),1,0)])
p <- upset(x, queries = list(list(query=elements,params=list("reg","Up"),color="red",active=T)), order.by = "freq",  nsets = n)
png(paste0(results.dir,"GSEA.scRNAseq.combined.png"),width = 1500,height = 1000,res=300)
p
dev.off()
```


# Combined

## Model: ~ 0 + pdx + cohort + sequencing + trial + mous

Let us look at the tissue types individually to see if they react differently.

```{r}
results.dir <- paste0(project.dir,"results/PARTNER/combined/")
invisible(lapply(grep("dir$",ls(),value = T), function(dname) dir.create(eval(as.name(dname)),recursive = T,showWarnings = F)))

x <- subset.PDX(rnaseq, rnaseq$meta[,Cohort %in% c("Untreated","Olaparib","AZD1775","Olap.AZD1775") & Sampling=="PT"])

cohort <- droplevels(x$meta[,Cohort])
trial <- factor(x$meta$Trial)
trial[is.na(trial)] <- "ABT008"
pdx <- factor(x$meta$project)
mouse <- factor(x$meta$Mouse)
sequencing <- factor(x$meta$merged.lanes)

d <- model.matrix(~ 0 + pdx + cohort + sequencing + trial + mouse)
d <- d[,colSums(d)>1]
d <- d[,-which(colnames(d)=="sequencing5")]
d

cont <- 3:5
names(cont) <- gsub("cohort","",colnames(d)[cont])
n <- length(cont)

DE.data <- DE.run(x$human$data,d,coefs = cont)

#Tables of DE genes
for(i in 1:n){
  DE <- topTable.annotated.DE(DE.data$efit,coef = i)
  fwrite(DE,paste0(results.dir,"DE.genes.",names(cont)[i],".csv"))
}

#Overlap of DE genes
DE <- decideTest.annotated.DE(DE.data$efit,p.val = 0.05)
fwrite(DE[.N:1],paste0(results.dir,"DE.genes.combined.csv"))

#Overlap plot
if(nrow(DE)){
  x <- rbind(
    cbind(DE[,ifelse(.SD[,2:(n+1)]>0,1,0)],data.table(reg="UP")),
    cbind(DE[,ifelse(.SD[,2:(n+1)]<0,1,0)],data.table(reg="DOWN"))
  )
  p <- upset(
    x,
    queries = list(list(query=elements,params=list("reg","UP"),color="red",active=T)),
    order.by = "freq",
    nsets = n,
    nintersects = 20
  )
  png(paste0(results.dir,"DE.genes.combined.png"),width = 1500,height = 1000,res=300)
  p
  dev.off()
}

#Gene set 
hm.gs <- load_gene_set(paste0(root.dir,"shared.data/gene.lists/h.all.v6.2.entrez.gmt"))
x <- lapply(1:n,function(i) GSEA.run(DE.data$v,d,cont[i],hm.gs,0.05,igc=0.05))
names(x) <- names(cont)
for(i in 1:n) fwrite(x[[i]],paste0(results.dir,"GSEA.hallmark.",names(cont)[i],".csv"))

#Gene set overlap
x <- rbindlist(x,idcol = "contrast")
x[,rank:=1:.N,contrast]
x <-  rbind(
  cbind(dcast(x[Direction=="Up"],pathway~contrast,value.var = "rank"),data.table(reg="Up")),
  cbind(dcast(x[Direction=="Down"],pathway~contrast,value.var = "rank"),data.table(reg="Down")),
  fill=TRUE
)
x <- cbind(x[,-grep("reg",colnames(x)),with=F],x[,.(reg)])
x[,avg.rank:=rowMeans(.SD[,2:(ncol(x)-1)],na.rm = T)]
fwrite(x[order(avg.rank)],paste0(results.dir,"GSEA.hallmark.combined.csv"))

#Overlap plot
x <- cbind(x[,.(reg)],x[,ifelse(!is.na(.SD[,2:(ncol(x)-2)]),1,0)])
p <- upset(x, queries = list(list(query=elements,params=list("reg","Up"),color="red",active=T)), order.by = "freq",  nsets = n)
png(paste0(results.dir,"GSEA.hallmark.combined.png"),width = 1500,height = 1000,res=300)
p
dev.off()
```

# Save

```{r}
results.dir <- paste0(project.dir,"results/PARTNER/")
save.image(file=paste0(results.dir,'myEnvironment.RData'))
```
