---
title: 'PDX: COX2-Stroma pathway'
author: "Alistair Martin"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output:
  html_document:
    toc: yes
    toc_float: yes
    code_folding: hide
layout: page
---


#Load libraries and data

```{r, message=FALSE, warning=FALSE}
root.dir <- "~/OneDrive/projects/"
source(paste0(root.dir,"R.utils/RNASEQ.utils.R"))

project.dir <- paste0(root.dir,"PDX/")
source(paste0(project.dir,"src/load.PDX.R"))
rnaseq <- get.PDX()
meta <- load.meta()

meta <- subset.meta(meta,meta[,project=="characterisation"])
rnaseq <- subset.PDX(rnaseq, rnaseq$meta[,Sample.name] %in% meta[,Sample.name])
rnaseq <- merge.PDX(rnaseq)
rnaseq$meta[,merged.lanes:=as.numeric(as.factor(merged.lanes))]

rnaseq$meta <- merge(rnaseq$meta,meta,sort=F)
rnaseq <- subset.PDX(rnaseq, rnaseq$meta[,sample.type == "xenograft"])
rnaseq <- subset.PDX(rnaseq, rnaseq$meta[,human.library.size > 1e6])
rnaseq <- subset.PDX(rnaseq, rnaseq$meta[,mouse.library.size/(human.library.size+mouse.library.size) < .8])
```

```{r}
pca.human <- pca.run(rnaseq$human)
pca.human <-cbind(pca.human,rnaseq$meta)
ggpairs(pca.human,columns = 1:3,mapping = aes(color=factor(merged.lanes)),title="Coloured by sequencing runs",progress = F)
ggpairs(pca.human,columns = 1:3,mapping = aes(color=factor(patient)),title="Coloured by patient",progress = F)
ggpairs(pca.human,columns = 1:3,mapping = aes(color=cut(mouse.library.size/(human.library.size+mouse.library.size),quantile(mouse.library.size/(human.library.size+mouse.library.size)))),title="Coloured by mouse fraction",progress = F)
ggpairs(pca.human,columns = 1:3,mapping = aes(color=(mouse.library.size/(human.library.size+mouse.library.size)>0.02)),title="Coloured by mouse fraction > 2%",progress = F)
```

```{R}
pca.human[,.(Sample.name,mouse.library.size,human.library.size,mouse.library.size/(human.library.size+mouse.library.size))][order(V4)]
```

Fraction of mouse/human has a clear effect on the human transcriptome.


```{r}
rnaseq.subset <- subset.PDX(rnaseq, rnaseq$meta[,mouse.library.size > 1e5])
pca.mouse <- pca.run(rnaseq.subset$mouse)
pca.mouse <-cbind(pca.mouse,rnaseq$meta[mouse.library.size > 1e5])
ggpairs(pca.mouse,columns = 1:3,mapping = aes(color=factor(merged.lanes)),title="Coloured by sequencing runs",progress = F)
ggpairs(pca.mouse,columns = 1:3,mapping = aes(color=factor(patient)),title="Coloured by patient",progress = F)
ggpairs(pca.mouse,columns = 1:3,mapping = aes(color=cut(mouse.library.size/(human.library.size+mouse.library.size),quantile(mouse.library.size/(human.library.size+mouse.library.size)))),title="Coloured by mouse fraction",progress = F)
```


# Signature w.r.t human reads

```{r}
PoI <- list(
  "immuno.markers" = "PDGFRB FAP ITGB1 ACTA2 S100A4 CAV1"
)
PoI <- sapply(PoI, strsplit, split=" ")
PoI <- lapply(PoI, function(x) data.table(biomaRt::select(org.Hs.eg.db,keys=x,keytype = "SYMBOL",columns = c("ENSEMBL","ENTREZID","SYMBOL"))))

x <- norm.count.matrix(rnaseq$human[rowSums(rnaseq$human)!=0,])
x <- x[match(PoI$immuno.markers$ENSEMBL,rownames(x)),]
rownames(x) <- PoI$immuno.markers$SYMBOL
pheatmap(x,main="Human")
```

```{r}
firstup <- function(x){
  x <- tolower(x)
  substr(x, 1, 1) <- toupper(substr(x, 1, 1))
  return(x)
}
PoI <- list(
  "immuno.markers" = "PDGFRB FAP ITGB1 ACTA2 S100A4 CAV1"
)
PoI <- sapply(PoI, strsplit, split=" ")
PoI <- lapply(PoI, function(p) sapply(p,firstup))
PoI <- lapply(PoI, function(x) data.table(biomaRt::select(org.Mm.eg.db,keys=x,keytype = "SYMBOL",columns = c("ENSEMBL","ENTREZID","SYMBOL"))))

x <- norm.count.matrix(subset.PDX(rnaseq, rnaseq$meta[,mouse.library.size > 1e5])$mouse)
x <- x[match(PoI$immuno.markers$ENSEMBL,rownames(x)),]
rownames(x) <- PoI$immuno.markers$SYMBOL
pheatmap(x,main="Mousex")
```